crates/bevy_color/src/color_gradient.rs

```diff
@@ -0,0 +1,104 @@
+use crate::Mix;
+use bevy_math::curve::{cores::EvenCoreError, Curve, Interval, SampleCurve};
+
+/// A curve whose samples are defined by a collection of colors.
+#[derive(Clone, Debug)]
+pub struct ColorCurve<T: Mix + Clone, I> {
+    curve: SampleCurve<T, I>,
+}
+
+impl<T, I> ColorCurve<T, I>
+where
+    T: Mix + Clone,
+{
+    /// Create a new [`ColorCurve`] from a collection of [mixable] types and a mixing function. The
+    /// domain of this curve will always be `[0.0, len - 1]` where `len` is the amount of mixable
+    /// objects in the collection.
+    ///
+    /// This fails if there's not at least two mixable things in the collection.
+    ///
+    /// [mixable]: `Mix`
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use bevy_color::palettes::basic::*;
+    /// # use bevy_color::Mix;
+    /// # use bevy_color::Srgba;
+    /// # use bevy_color::ColorCurve;
+    /// # use bevy_math::curve::Interval;
+    /// # use bevy_math::curve::Curve;
+    /// let broken = ColorCurve::new([RED], Srgba::mix);
+    /// assert!(broken.is_err());
+    /// let gradient = ColorCurve::new([RED, GREEN, BLUE], Srgba::mix);
+    /// assert!(gradient.is_ok());
+    /// assert_eq!(gradient.unwrap().domain(), Interval::new(0.0, 2.0).unwrap());
+    /// ```
+    pub fn new(
+        colors: impl IntoIterator<Item = impl Into<T>>,
+        interpolation: I,
+    ) -> Result<Self, EvenCoreError>
+    where
+        I: Fn(&T, &T, f32) -> T,
+    {
+        let colors = colors.into_iter().map(Into::into).collect::<Vec<_>>();
+        Interval::new(0.0, colors.len().saturating_sub(1) as f32)
+            .map_err(|_| EvenCoreError::NotEnoughSamples {
+                samples: colors.len(),
+            })
+            .and_then(|domain| SampleCurve::new(domain, colors, interpolation))
+            .map(|curve| Self { curve })
+    }
```

Comment:
```suggestion
    pub fn new(colors: impl IntoIterator<Item = T>) -> Result<Self, EvenCoreError> {
        let mix: fn(&T, &T, f32) -> T = <T as Mix>::mix;
        let colors = colors.into_iter().collect::<Vec<_>>();
        Interval::new(0.0, colors.len().saturating_sub(1) as f32)
            .map_err(|_| EvenCoreError::NotEnoughSamples {
                samples: colors.len(),
            })
            .and_then(|domain| SampleCurve::new(domain, colors, mix))
            .map(|curve| Self { curve })
    }
```

I've also removed the `impl Into<T>` from the `colours` parameter. The reason is an iterator can only ever be homogeneous on the type it stores, so all that generalisation does is let a user skip a `.map(Srgba::from)` line. But, that is at the expense of _always_ needing to specify the colour space you want the gradient in, when it's reasonable to assume you want it in the same colour space you're already in.